drivers/adc/adc_xmc4xxx.c

```diff
@@ -0,0 +1,313 @@
+
+#define ADC_CONTEXT_USES_KERNEL_TIMER
+#include "adc_context.h"
+
+#define LOG_LEVEL CONFIG_ADC_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(adc_xmc4xxx);
+
+struct adc_xmc4xxx_data {
+	struct adc_context ctx;
+	const struct device *dev;
+	uint16_t *buffer;
+	uint16_t *repeat_buffer;
+	uint8_t channel_mask;
+};
+
+struct adc_xmc4xxx_cfg {
+	XMC_VADC_GROUP_t *base;
+	void (*irq_cfg_func)(void);
+	uint8_t irq_num;
+};
+
+static int adc_global_init;
+XMC_VADC_GLOBAL_t *adc_global_ptr = (XMC_VADC_GLOBAL_t *)0x40004000;
+
+static void adc_context_start_sampling(struct adc_context *ctx)
+{
+	struct adc_xmc4xxx_data *data = CONTAINER_OF(ctx, struct adc_xmc4xxx_data, ctx);
+	const struct device *dev = data->dev;
+	const struct adc_xmc4xxx_cfg *config = dev->config;
+	VADC_G_TypeDef *adc_group = config->base;
+
+	data->repeat_buffer = data->buffer;
+
+	/* generate a load event */
+	adc_group->ASMR |= BIT(VADC_G_ASMR_LDEV_Pos);
+
+	XMC_VADC_GROUP_ScanEnableArbitrationSlot(adc_group);
+}
+
+static void adc_context_update_buffer_pointer(struct adc_context *ctx,
+					      bool repeat_sampling)
+{
+	struct adc_xmc4xxx_data *data = CONTAINER_OF(ctx, struct adc_xmc4xxx_data, ctx);
+
+	if (repeat_sampling) {
+		data->buffer = data->repeat_buffer;
+	}
+}
+
+static void adc_xmc4xxx_isr(const struct device *dev)
+{
+	struct adc_xmc4xxx_data *data = dev->data;
+	const struct adc_xmc4xxx_cfg *config = dev->config;
+	XMC_VADC_GROUP_t *adc_group = config->base;
+	uint32_t channel_mask = data->channel_mask;
+	uint32_t ch;
+
+	/* Conversion has completed. */
+	while (channel_mask > 0) {
+		ch = find_lsb_set(channel_mask) - 1;
+		*data->buffer++ = XMC_VADC_GROUP_GetResult(adc_group, ch);
+		channel_mask &= ~BIT(ch);
+	}
+
+	adc_context_on_sampling_done(&data->ctx, dev);
+	LOG_DBG("%s ISR triggered.", dev->name);
+}
+
+
+static int start_read(const struct device *dev,
+		      const struct adc_sequence *sequence)
+{
+	struct adc_xmc4xxx_data *data = dev->data;
+	const struct adc_xmc4xxx_cfg *config = dev->config;
+	XMC_VADC_GROUP_t *adc_group = config->base;
+	uint32_t requested_channels = sequence->channels;
+	uint8_t resolution = sequence->resolution;
+	uint32_t configured_channels = adc_group->ASSEL & requested_channels;
+
+	if (requested_channels == 0) {
+		LOG_ERR("No channels requested");
+		return -EINVAL;
+	}
+
+	if (requested_channels != configured_channels) {
+		LOG_ERR("Selected channels not configured");
+		return -EINVAL;
+	}
+
+	if (sequence->oversampling) {
+		LOG_ERR("oversampling not supported");
+		return -ENOTSUP;
+	}
+
+	if (resolution == 8) {
+		adc_group->ICLASS[0] |=  2 << VADC_G_ICLASS_CMS_Pos;
+	} else if (resolution == 10) {
+		adc_group->ICLASS[0] |=  1 << VADC_G_ICLASS_CMS_Pos;
+	} else if (resolution == 12) {
+		adc_group->ICLASS[0] |=  0 << VADC_G_ICLASS_CMS_Pos;
+	} else {
+		LOG_ERR("Invalid resolution");
+		return -EINVAL;
+	}
+
+	data->channel_mask = requested_channels;
+	data->buffer = sequence->buffer;
```

Comment:
It should be also checked if the buffer will fit results for all requested samples: https://github.com/zephyrproject-rtos/zephyr/blob/b7cc0cd05635a2e745d5ac3710f4cd0ff5fa4939/include/zephyr/drivers/adc.h#L518-L524

Reply:
thanks, I've added a check.